### PR TITLE
Add support for bearer tokens in Jenkins CLI

### PR DESCRIPTION
**Pull request by sorend**
See [JENKINS-61965](https://issues.jenkins-ci.org/browse/JENKINS-61965).

### Proposed changelog entries

* Support for Bearer tokens in Jenkins-CLI -auth parameter.

### Proposed upgrade guidelines

N/A

### CODE REVIEW

File: cli/src/main/java/hudson/cli/CLIConnectionFactory.java [hunk 1]
```diff
@@ -36,4 +36,19 @@ public CLIConnectionFactory basicAuth(String userInfo) {
         return authorization("Basic " + Base64.getEncoder().encodeToString(userInfo.getBytes()));
     }
 
+    /**
+     * Convenience method to call {@link #authorization} with the HTTP bearer authentication.
+     * Cf. {@code BasicHeaderApiTokenAuthenticator}.
+     */
+    public CLIConnectionFactory bearerAuth(String bearerToken) {
+        return authorization("Bearer " + bearerToken);
```
Comment (sladyn98) [inline cli/src/main/java/hudson/cli/CLIConnectionFactory.java:44]: Is it possible that the bearer token here is empty ?

File: cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package hudson.cli;
+
+import org.junit.Assert;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class CLIConnectionFactoryTest {
+
+    CLIConnectionFactory sut;
```
Comment (sladyn98) [inline cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java:9]:
```suggestion
    CLIConnectionFactory cliFactory;
```

File: cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    @BeforeEach
+    void setUp() {
+        sut = new CLIConnectionFactory();
+    }
+
+    @Test
+    void testBearerFromToken() {
+        Assert.assertEquals("Bearer some-token", sut.of("some-token").authorization);
+    }
+
```
Comment (sladyn98) [inline cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java:20]: Would it be better if we also include  an Empty Bearer Token test ?

**Review by sladyn98 — CHANGES_REQUESTED** [review]

File: cli/src/main/java/hudson/cli/CLIConnectionFactory.java (hunk 1 above)
Comment (sorend) [reply cli/src/main/java/hudson/cli/CLIConnectionFactory.java:44]: From my understanding of CLI class, it can be an empty string but not null -- follows the semantics of basicAuth

File: cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java (hunk 3 above)
Comment (sorend) [reply cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java:20]: I can add the test, but the cli will behave in the same was as if empty auth is passed to basicAuth, semantics are the same. What do you suggest?

File: cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    @Test
+    void testBearerFromToken() {
+        Assert.assertEquals("Bearer some-token", cliFactory.of("some-token").authorization);
+    }
+
+    @Test
+    void testBearerFromEmptyToken() {
+        Assert.assertEquals("Bearer some-token", cliFactory.of("some-token").authorization);
```
Comment (varyvol) [inline cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java:23]: This is not empty when the test name suggests so?

Comment (sorend) [reply cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java]: Sorry a missing cleanup.

File: cli/src/main/resources/hudson/cli/client/Messages.properties [hunk 5]
```diff
@@ -11,10 +11,12 @@ CLI.Usage=Jenkins CLI\n\
   \ -user               : specify user (for use with -ssh)\n\
   \ -strictHostKey      : request strict host key checking (for use with -ssh)\n\
   \ -logger FINE        : enable detailed logging from the client\n\
-  \ -auth [ USER:SECRET | @FILE ] : specify username and either password or API token (or load from them both from a file);\n\
-  \                                 for use with -http.\n\
-  \                                 Passing credentials by file is recommended.\n\
-  \                                 See https://jenkins.io/redirect/cli-http-connection-mode for more info and options.\n\
+  \ -auth [ USER:SECRET | TOKEN | @FILE ]\n\
+  \                     : specify username and either password or API token (or load from them both from a file);\n\
```
Comment (sladyn98) [inline cli/src/main/resources/hudson/cli/client/Messages.properties:15]:
```suggestion
  \                     : specify username and  password or API token (or load from them both from a file);\n\
```

File: cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java (hunk 3 above)
Comment (sladyn98) [reply cli/src/test/java/hudson/cli/CLIConnectionFactoryTest.java:20]: Okay, I do not have a deep understanding of basicAuth here, if I understand correctly if empty token is passed it will be handled normally and therefore we do not need to test it. Might need some additional review
CC @daniel-beck

**Review by sladyn98 — CHANGES_REQUESTED** [review]

File: cli/src/main/resources/hudson/cli/client/Messages.properties (hunk 5 above)
Comment (sorend) [reply cli/src/main/resources/hudson/cli/client/Messages.properties:15]: I think this changes the meaning. I added the next line about bearer token. This is about that you can provide Jenkins API token or Jenkins password, so I think it should be left as is.

Comment (sladyn98) [reply cli/src/main/resources/hudson/cli/client/Messages.properties:15]: Okay thanks :+1:

**Review by varyvol — APPROVED** [review]

**Comment by daniel-beck** [comment]
Is this the only reasonable choice for single-element authentication, or are there other schemes competing with `Bearer` for getting this feature? ideally this behavior would hold up for several years.

**Comment by sorend** [comment]
@daniel-beck 
> Is this the only reasonable choice for single-element authentication, or are there other schemes competing with `Bearer` for getting this feature? ideally this behavior would hold up for several years.

There are other schemes yes, but I think it has to be seen in the context of what types of authentication Jenkins can accept. This should work with setups using OAuth2 backed login, such as the Openshift Login (as in my case), Google Login, and Github Login.

**Review by fcojfernandez — APPROVED** [review]

**Review by alecharp — APPROVED** [review]

**Comment by varyvol** [comment]
@sorend I already approved the PR, but what @daniel-beck mentions about the future make sense. Given you seem to agree, or at least think it's reasonable, maybe you could work on implementing that?

**Comment by MRamonLeon** [comment]
@sorend could the token have a colon **:** ?

I'm with Daniel, something looking forward to new methods would be great, but it can also be tackled on another PR.

Still the question is relevant

**Comment by sorend** [comment]
@varyvol -- sure, if the general opinion prefers a separate parameter, then I will implement it.

@MRamonLeon -- no, the token can not contain a colon (according to the RFC: https://tools.ietf.org/html/rfc6750#page-5 )

**Comment by sorend** [comment]
@daniel-beck @varyvol @MRamonLeon -- I have updated the PR, so jenkins-cli now has an additional (optional) -bearer parameter.

File: cli/src/main/resources/hudson/cli/client/Messages.properties [hunk 6]
```diff
@@ -15,6 +15,7 @@ CLI.Usage=Jenkins CLI\n\
   \                                 for use with -http.\n\
   \                                 Passing credentials by file is recommended.\n\
   \                                 See https://jenkins.io/redirect/cli-http-connection-mode for more info and options.\n\
+  \ -bearer [ TOKEN | @FILE ]     : specify authentication using a bearer token (or load the token from file)\n\
```
Comment (MRamonLeon) [inline cli/src/main/resources/hudson/cli/client/Messages.properties:18]: it is exclusive to `-auth`. Can we check it and say it explicitly on the help. Also `-bearer` is for -http I guess, and passing it with a file also applies.

Sorry for being picky, I was about to approve when you said it cannot have a :smile: but with the new way, I believe is worth to document it well.

Comment (sorend) [reply cli/src/main/resources/hudson/cli/client/Messages.properties]: @MRamonLeon Thank you for the review :+1:  I have updated the message and added a warning when both are used (this is how -user and -auth exclusivity was handled already)

**Review by MRamonLeon — APPROVED** [review]

**Comment by timja** [comment]
We'll merge this in 24 hours if no negative feedback

**Review by oleg-nenashev — APPROVED** [review]
LGTM. Will proceed with the merge.
Thanks for your contribution @sorend !
